File: crates/prolly/bindings/wasm/examples/basic-map.ts
import assert from "node:assert/strict";
import { existsSync, readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

import { loadProllyWasm } from "../src/index.ts";

const here = dirname(fileURLToPath(import.meta.url));
const pkgWasm = join(here, "../pkg/prolly_wasm_bg.wasm");
if (!existsSync(pkgWasm)) throw new Error("Run `npm --prefix crates/prolly/bindings/wasm run build:wasm` first.");

const wasm = await loadProllyWasm("../pkg/prolly_wasm.js", readFileSync(pkgWasm));
const bytes = (value: string): Uint8Array => new TextEncoder().encode(value);
const text = (value: Uint8Array | null | undefined): string | null =>
  value == null ? null : new TextDecoder().decode(value);

const engine = wasm.WasmProllyEngine.memory();
let tree = engine.create();
tree = engine.put(tree, bytes("user:001"), bytes("Ada"));
tree = engine.put(tree, bytes("user:002"), bytes("Grace"));
tree = engine.put(tree, bytes("user:003"), bytes("Linus"));

assert.equal(text(engine.get(tree, bytes("user:001"))), "Ada");
tree = engine.delete(tree, bytes("user:003"));
assert.equal(engine.get(tree, bytes("user:003")), null);

const users = engine.range(tree, bytes("user:"), bytes("user;"));
assert.equal(users.length, 2);

console.log(`basic_map: ${users.length} users in range`);

File: crates/prolly/examples/basic_map.rs
use prolly::{Config, MemStore, Prolly};

fn main() -> Result<(), prolly::Error> {
    let prolly = Prolly::new(MemStore::new(), Config::default());

    let tree = prolly.create();
    let tree = prolly.put(&tree, b"user:001".to_vec(), b"Ada".to_vec())?;
    let tree = prolly.put(&tree, b"user:002".to_vec(), b"Grace".to_vec())?;
    let tree = prolly.put(&tree, b"user:003".to_vec(), b"Linus".to_vec())?;

    assert_eq!(prolly.get(&tree, b"user:001")?, Some(b"Ada".to_vec()));

    let tree = prolly.delete(&tree, b"user:003")?;
    assert_eq!(prolly.get(&tree, b"user:003")?, None);

    let users = prolly
        .range(&tree, b"user:", Some(b"user;"))?
        .collect::<Result<Vec<_>, _>>()?;

    assert_eq!(
        users,
        vec![
            (b"user:001".to_vec(), b"Ada".to_vec()),
            (b"user:002".to_vec(), b"Grace".to_vec()),
        ]
    );

    println!("{} users in range", users.len());
    Ok(())
}
